ext/pdo_sqlite/pdo_sqlite.c

```diff
@@ -21,39 +21,654 @@
+{
+	php_stream_pdosqlite3_data *sqlite3_stream = (php_stream_pdosqlite3_data *) stream->abstract;
+	ssb->sb.st_size = sqlite3_stream->size;
+	return 0;
+}
+
+static const php_stream_ops php_stream_pdosqlite3_ops = {
+		php_pdosqlite3_stream_write,
+		php_pdosqlite3_stream_read,
+		php_pdosqlite3_stream_close,
+		php_pdosqlite3_stream_flush,
+		"PDOSQLite",
+		php_pdosqlite3_stream_seek,
+		php_pdosqlite3_stream_cast,
+		php_pdosqlite3_stream_stat,
+		NULL
+};
+
+/* Open a blob as a stream which we can read / write to. */
+PHP_METHOD(PdoSqlite, openBlob)
+{
+	char *table, *column, *dbname = "main", *mode = "rb";
+	size_t table_len, column_len, dbname_len;
+	zend_long rowid, flags = SQLITE_OPEN_READONLY, sqlite_flags = 0;
+	sqlite3_blob *blob = NULL;
+	php_stream_pdosqlite3_data *sqlite3_stream;
+	php_stream *stream;
+
+	pdo_dbh_t *dbh;
+	pdo_sqlite_db_handle *db_handle;
+
+	dbh = Z_PDO_DBH_P(ZEND_THIS);
+	PDO_CONSTRUCT_CHECK;
+	db_handle = (pdo_sqlite_db_handle *)dbh->driver_data;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "ssl|pl", &table, &table_len, &column, &column_len, &rowid, &dbname, &dbname_len, &flags) == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	sqlite3 *sqlite_handle;
+	sqlite_handle = db_handle->db;
+	sqlite_flags = (flags & SQLITE_OPEN_READWRITE) ? 1 : 0;
+
+	if (sqlite3_blob_open(sqlite_handle, dbname, table, column, rowid, sqlite_flags, &blob) != SQLITE_OK) {
+		// TODO - exception needed here.
+//		php_sqlite3_error(db_obj, "Unable to open blob: %s", sqlite3_errmsg(db_obj->db));
+		// and converting to exception zend_throw_exception_ex(php_pdo_get_exception(), 0, errtext);
+		RETURN_FALSE;
+	}
+
+	sqlite3_stream = emalloc(sizeof(php_stream_pdosqlite3_data));
+	sqlite3_stream->blob = blob;
+	sqlite3_stream->flags = flags;
+	sqlite3_stream->position = 0;
+	sqlite3_stream->size = sqlite3_blob_bytes(blob);
+
+	if (sqlite_flags != 0) {
+		mode = "r+b";
+	}
+
+	stream = php_stream_alloc(&php_stream_pdosqlite3_ops, sqlite3_stream, 0, mode);
+
+	if (stream) {
+		php_stream_to_zval(stream, return_value);
+	} else {
+		RETURN_FALSE;
+	}
+}
+
+static int php_pgsql_collation_callback(void *context, int string1_len, const void *string1,
+	int string2_len, const void *string2)
+{
+	int ret;
+	zval zargs[2];
+	zval retval;
+	struct pdo_sqlite_collation *collation = (struct pdo_sqlite_collation*) context;
+
+	collation->fc.fci.size = sizeof(collation->fc.fci);
+	ZVAL_COPY_VALUE(&collation->fc.fci.function_name, &collation->callback);
+	collation->fc.fci.object = NULL;
+	collation->fc.fci.retval = &retval;
+
+	// Prepare the arguments.
+	ZVAL_STRINGL(&zargs[0], (char *) string1, string1_len);
+	ZVAL_STRINGL(&zargs[1], (char *) string2, string2_len);
+	collation->fc.fci.param_count = 2;
+	collation->fc.fci.params = zargs;
+
+	if ((ret = zend_call_function(&collation->fc.fci, &collation->fc.fcc)) == FAILURE) {
+		php_error_docref(NULL, E_WARNING, "An error occurred while invoking the callback");
```

Comment:
I'm not sure this is possible to happen?

Reply:
Why not?

Reply:
Well how? By trying to call an undefined function in SQL? That will probably fail on the DB level, if a function has been registered than the FCC is stored and is always possible to be called.

I'm pretty sure this could use the more modern FCC API too, but I can refactor that later.

Reply:
Those this have the trampoline tests like sqlite3 has?